File: Core/src/main/java/com/braintreepayments/api/internal/TLSSocketFactory.java
package com.braintreepayments.api.internal;

import java.io.IOException;
import java.net.InetAddress;
import java.net.Socket;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;

import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLException;
import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;

class TLSSocketFactory extends SSLSocketFactory {

    protected SSLSocketFactory mInternalSSLSocketFactory;

    protected TLSSocketFactory() throws SSLException {
        try {
            SSLContext sslContext = SSLContext.getInstance("TLS");
            sslContext.init(null, null, null); // use system security providers
            mInternalSSLSocketFactory = sslContext.getSocketFactory();
        } catch (NoSuchAlgorithmException | KeyManagementException e) {
            throw new SSLException(e.getMessage());
        }
    }

    @Override
    public String[] getDefaultCipherSuites() {
        return mInternalSSLSocketFactory.getDefaultCipherSuites();
    }

    @Override
    public String[] getSupportedCipherSuites() {
        return mInternalSSLSocketFactory.getSupportedCipherSuites();
    }

    @Override
    public Socket createSocket(Socket s, String host, int port, boolean autoClose)
            throws IOException {
        return enableTLSOnSocket(mInternalSSLSocketFactory.createSocket(s, host, port, autoClose));
    }

    @Override
    public Socket createSocket(String host, int port) throws IOException {
        return enableTLSOnSocket(mInternalSSLSocketFactory.createSocket(host, port));
    }

    @Override
    public Socket createSocket(String host, int port, InetAddress localHost, int localPort)
            throws IOException {
        return enableTLSOnSocket(
                mInternalSSLSocketFactory.createSocket(host, port, localHost, localPort));
    }

    @Override
    public Socket createSocket(InetAddress host, int port) throws IOException {
        return enableTLSOnSocket(mInternalSSLSocketFactory.createSocket(host, port));
    }

    @Override
    public Socket createSocket(InetAddress address, int port, InetAddress localAddress,
            int localPort) throws IOException {
        return enableTLSOnSocket(
                mInternalSSLSocketFactory.createSocket(address, port, localAddress, localPort));
    }

    private Socket enableTLSOnSocket(Socket socket) {
        if (socket instanceof SSLSocket) {
            ArrayList<String> supportedProtocols =
                    new ArrayList<>(Arrays.asList(((SSLSocket) socket).getSupportedProtocols()));
            supportedProtocols.retainAll(Arrays.asList("TLSv1.2", "TLSv1.1", "TLSv1"));

            ((SSLSocket)socket).setEnabledProtocols(supportedProtocols.toArray(
                    new String[supportedProtocols.size()]));
        }

        return socket;
    }
}

File: SharedUtils/src/main/java/com/braintreepayments/api/sharedutils/TLSSocketFactory.kt
package com.braintreepayments.api.sharedutils

import androidx.annotation.RestrictTo
import java.io.IOException
import java.io.InputStream
import java.net.InetAddress
import java.net.Socket
import java.security.KeyStore
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import javax.net.ssl.SSLContext
import javax.net.ssl.SSLException
import javax.net.ssl.SSLSocket
import javax.net.ssl.SSLSocketFactory
import javax.net.ssl.TrustManagerFactory

/**
 * `TLSSocketFactory` is a custom implementation of [SSLSocketFactory] that enforces the use of specific TLS protocols
 * (TLSv1.2 and TLSv1.3) and allows for the management of trusted certificates from a provided [InputStream].
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class TLSSocketFactory(
    certificateStream: InputStream,
    keyStore: KeyStore = KeyStore.getInstance(KeyStore.getDefaultType()),
    certificateFactory: CertificateFactory = CertificateFactory.getInstance("X.509"),
    trustManagerFactory: TrustManagerFactory = TrustManagerFactory.getInstance(
        TrustManagerFactory.getDefaultAlgorithm()
    ),
    sslContext: SSLContext = SSLContext.getInstance("TLS")
) : SSLSocketFactory() {

    private val internalSSLSocketFactory: SSLSocketFactory

    init {
        try {
            keyStore.load(null, null)

            val certificates = certificateFactory.generateCertificates(certificateStream)
            for (cert in certificates) {
                if (cert is X509Certificate) {
                    keyStore.setCertificateEntry(cert.subjectDN.name, cert)
                }
            }

            trustManagerFactory.init(keyStore)
            sslContext.init(null, trustManagerFactory.trustManagers, null)
            internalSSLSocketFactory = sslContext.socketFactory
        } catch (e: Exception) {
            throw SSLException(e.message)
        } finally {
            try {
                certificateStream.close()
            } catch (ignored: IOException) {
            } catch (ignored: NullPointerException) {
            }
        }
    }

    override fun getDefaultCipherSuites(): Array<String> {
        return internalSSLSocketFactory.defaultCipherSuites
    }

    override fun getSupportedCipherSuites(): Array<String> {
        return internalSSLSocketFactory.supportedCipherSuites
    }

    @Throws(IOException::class)
    override fun createSocket(
        s: Socket,
        host: String,
        port: Int,
        autoClose: Boolean
    ): Socket {
        return enableTLSOnSocket(internalSSLSocketFactory.createSocket(s, host, port, autoClose))
    }

    @Throws(IOException::class)
    override fun createSocket(host: String, port: Int): Socket {
        return enableTLSOnSocket(internalSSLSocketFactory.createSocket(host, port))
    }

    @Throws(IOException::class)
    override fun createSocket(
        host: String,
        port: Int,
        localHost: InetAddress,
        localPort: Int
    ): Socket {
        return enableTLSOnSocket(
            internalSSLSocketFactory.createSocket(host, port, localHost, localPort)
        )
    }

    @Throws(IOException::class)
    override fun createSocket(host: InetAddress, port: Int): Socket {
        return enableTLSOnSocket(internalSSLSocketFactory.createSocket(host, port))
    }

    @Throws(IOException::class)
    override fun createSocket(
        address: InetAddress,
        port: Int,
        localAddress: InetAddress,
        localPort: Int
    ): Socket {
        return enableTLSOnSocket(
            internalSSLSocketFactory.createSocket(address, port, localAddress, localPort)
        )
    }

    private fun enableTLSOnSocket(socket: Socket): Socket {
        if (socket is SSLSocket) {
            val supportedProtocols = socket.supportedProtocols.toMutableList()
            supportedProtocols.retainAll(listOf("TLSv1.2", "TLSv1.3"))

            socket.enabledProtocols = supportedProtocols.toTypedArray()
        }

        return socket
    }
}
